Sources/SourceKit/clangd/ClangLanguageServer.swift

```diff
@@ -17,29 +17,136 @@ import LSPLogging
+  private var clangd: Connection!
 
-  var capabilities: ServerCapabilities? = nil
+  private var capabilities: ServerCapabilities? = nil
 
-  let buildSystem: BuildSystem
+  private let buildSystem: BuildSystem
 
-  let clang: AbsolutePath?
+  private let clangdPath: AbsolutePath
 
-  /// Creates a language server for the given client using the sourcekitd dylib at the specified path.
-  public init(client: Connection, clangd: Connection, buildSystem: BuildSystem,
-              clang: AbsolutePath?) throws {
-    self.clangd = clangd
-    self.buildSystem = buildSystem
-    self.clang = clang
+  private let clangdOptions: [String]
+
+  private let client: MessageHandler
+
+  private var state: LanguageServerState {
+     didSet {
+       for handler in stateChangeHandlers {
+         handler(oldValue, state)
+       }
+     }
+   }
+
+  private var stateChangeHandlers: [(_ oldState: LanguageServerState, _ newState: LanguageServerState) -> Void] = []
+
+  /// Ask the parent of this language server to re-open all documents in this language server.
+  private let reopenDocuments: () -> Void
+
+  /// The `InitializeRequest` with which `clangd` was originally initialized.
+  /// Stored so we can replay the initialization when clangd crashes.
+  private var initializeRequest: InitializeRequest?
+
+  public init(client: MessageHandler,
+              clangdPath: AbsolutePath,
+              buildSettings: BuildSystem?,
+              clangdOptions: [String],
+              reopenDocuments: @escaping () -> Void
+  ) throws {
+    self.client = client
+    self.buildSystem = buildSettings ?? BuildSystemList()
+    self.clangdPath = clangdPath
+    self.clangdOptions = clangdOptions
+    self.state = .connected
+    self.reopenDocuments = reopenDocuments
+  }
+
+  func initialize() throws {
+    let clientToServer: Pipe = Pipe()
+    let serverToClient: Pipe = Pipe()
+
+    let connection = JSONRPCConnection(
+      protocol: MessageRegistry.lspProtocol,
+      inFD: serverToClient.fileHandleForReading.fileDescriptor,
+      outFD: clientToServer.fileHandleForWriting.fileDescriptor
+    )
+
+    self.clangd = connection
+
+    connection.start(receiveHandler: client)
+
+    let process = Foundation.Process()
+
+    if #available(OSX 10.13, *) {
+      process.executableURL = clangdPath.asURL
+    } else {
+      process.launchPath = clangdPath.pathString
+    }
+
+    process.arguments = [
+      "-compile_args_from=lsp",   // Provide compiler args programmatically.
+      "-background-index=false",  // Disable clangd indexing, we use the build
+      "-index=false",             // system index store instead.
+    ] + clangdOptions
+
+    process.standardOutput = serverToClient
+    process.standardInput = clientToServer
+    process.terminationHandler = { [weak self] process in
+      log("clangd exited: \(process.terminationReason) \(process.terminationStatus)")
+      connection.close()
+      if process.terminationStatus != 0 {
```

Comment:
We’re already doing it that way, I just copied the code over from here. AFAIK it never hurts to close the connection, if it’s already closed, its a no-op. So it’s better to just close it.
https://github.com/apple/sourcekit-lsp/blob/master/Sources/SourceKit/clangd/ClangLanguageServer.swift#L238

Reply:
Ah, I meant the part below, where `clangd` is restarted. Maybe you should double check the `state` or see if it was shutting down or equivalent?

Reply:
What do you want to check. `state` is always either `connected` or `connectionInterrupted` in `ClangLanguageServerShim ` since we never shut the connection to `clangd` down. What kind of check are you proposing?

Reply:
We should probably handle the exit notification and properly shut it down, perhaps that should be in a follow up change though

Reply:
If we send an exit notification, I think it is safe to assume that `clangd` will exit with exit code 0 which means we won’t restart it (since we check for `process.terminationStatus != 0`). I think we are misunderstanding each other though.

Reply:
Ah I was thinking about signals, but I don't think SourceKit-LSP has any signal handlers right now so it should crash.

Reply:
Ah, OK. No, there’s no signal handlers so we should be fine.